### ISSUE TITLE
Stack overflow prevention

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#297]
A few years ago I found a nice article: https://www.embeddedrelated.com/showarticle/1574.php

I even posted it to our ML back then: http://mikro.naprvyraz.sk/mint/201402/threads.html#00086 and there weren't any serious objections so maybe if someone is bored enough, it could be done?

I really like this approach, way better than random application freezes (hello huge C++ projects...). It could be configurable, same way as we have e.g. `TPA_INITIALMEM` for process.

**Comment by th-otto** [comment]
I don't think that this approach would work for our target. If i understand it correctly, the idea is to put the stack below the data segment. In our case that would essentially mean swapping the data & bss segments. That would mean on a stack overflow the stack would run into the text segment instead of the data segment, which is even worse.

Also, it cannot simple be done by a linker script, because the stack is not part of the initial bss segment, but configurable at runtime (through stack utility, STACKSIZE env etc.) and may even depend on the amount of memory (if set to use 3/4 of available mem etc).

But gcc contains a libssp for quite some time (atleast 4.6.4 already has it). IiRC it emits some calls to this library when you compile with -fstack-check Maybe that just needs to be implement for 68k?

**Comment by czietz** [comment]
Apologies if I simply didn't get the idea correctly.

But to me, it sounds like a lot of effort for quite a small benefit. Only newly compiled software using a matching runtime library / startup code would be able to profit, as it's the application who sets its stack pointer (and not the OS) to the top of the memory. Hence, you would have to maintain two code paths in MiNT (for old and new applications) and two code paths in the runtime library (for OS with and without the feature).

**Comment by th-otto** [comment]
Actually, its both the OS and the startup code that sets the SP. The OS has to provide an initial stack where it (amongst others) passes the BASEPAGE pointer.

Not to mention that Mshrink is also used by programs that are not aware of that behaviour (eg. when loading modules). 

And then you won't have any protection when running such programs in TOS.

I totally agree with @czietz , lots of efforts with a small benefit. There are certainly more important things to work on.

**Comment by mikrosk** [comment]
I'd argue about the small benefit: for me this is like "daily" situation: I port some software (usually SDL-based but also gcc for instance), run it and it hangs. So the first thing I have to do is to increase stack size to some ridiculous value (like 1 MB) to see whether the default 64k isn't too small. That usually triggers the TPA limit so I have to change it in mint.cnf, reboot and try again. Really annoying.

This gets even more annoying when the program does _not_ freeze, I saw cases of weird behaviour attributed solely to the fact that the stack pointer was overwriting some non-essential data (like screen).

From my experience, very few people are aware of this problem (along with the need to use `-S` flag for apps using their own interrupts, btw) so it's very hard to debug and I have to repeat this hint over and over again.

You're basically correct but the two code paths:

- new OS + new mshrink: ok
- new OS + old mshrink: OS detects missing prg flag, allocates program TPA at the bottom of the memory block as before
- old OS + old mshrink: ok
- old OS / TOS + new mshrink: app would need to check for mint presence and its version/build date

Of course, for normal applications (games, demos, small apps) there's no reason for this (so no need to change anyone's workflow) as it's targeted for stack-hungry apps from Linux world and perhaps some debugging.

**Comment by czietz** [comment]
In a modern OS, if I wanted to catch stack overflow I would add a guard page between BSS and stack.

 

Otoh, I don't know what the granularity of the MMUs of the different 68k CPUs is, though - important because without virtual memory the guard page wastes actual memory. Neither do I know whether MiNT would already offer a function to place a write-protected memory page between BSS and stack. (That still needs a runtime library that sets the guard area up, but imho is much less invasive than reordering the memory layout.)

**Comment by mikrosk** [comment]
IIRC, it is 4k for 030 and whatever size it is already set for 040+ (usually 4k, too). So 4k per process isn't that bad.

Your idea isn't bad either but @th-otto correct me if I'm wrong here - it would also require some app support code - instead of stack size you'd need to allocate stack size + 4k to have that one guarded page to waste and the OS would need to know whether this is the intention (i.e. whether I meant to allocate 64k or 60k + the guard page). But maybe not, we could just do it for apps with stack size >= 64k automatically.

So yeah, FreeMiNT could do (perhaps optionally) some additional work here, just silently stealing 4k from the stack, mark it as protected and let the app do its stuff. Hmm.

**Comment by th-otto** [comment]
>   Otoh, I don't know what the granularity of the MMUs of the different 68k CPUs is, though

On 030, that can be any power of 2 between 256 bytes and 32k. On 040, only 4k and 8k is possible. MiNT internally use 8k memory blocks, even if the MMU uses 4k pages, so that would be the granularity used here.

> it would also require some app support code

Yes, but this would be much simpler. Only thing you have to check for is whether OS supports MP.

Only problem is the fact it has to be done at runtime rather than by the OS, so we would need something like mprotect().

>FreeMiNT could do (perhaps optionally) some additional work here, just silently stealing 4k from the stack

That won't work, because freemint does not know about the stacksize the app is going to use.

I would still prefer some simpler solution, like emitting runtime checks on function entry or similar. Even Pure-C can do that.

**Comment by mikrosk** [comment]
> That won't work, because freemint does not know about the stacksize the app is going to use.

Well, it knows the basepage of the process. And it can intercept Mshrink() with that address. So it would need some pairing.

> I would still prefer some simpler solution, like emitting runtime checks on function entry or similar.

Someone mentioned that MagiC is using canaries like that but it has a performance hit. What the previous two solutions don't.

**Comment by skaftetryne** [comment]
> Well, it knows the basepage of the process. And it can intercept Mshrink() with that address. So it would need some pairing.

I much prefer the slight performance hit of runtime checks over this kind of manipulation. The user stack is 100% the application's responsibility, I don't think the OS should interfere with that.

What about a special Mxalloc mode to allocate stack space? This could leave a guard page when MP is active. This will only work with applications compiled with support for this, but since stack checking is primarily a debug tool that wouldn't be a problem IMO.

**Comment by th-otto** [comment]
I took a brief look at it now to implement something in gcc. Basically, there are already two mechanism:

- -fstack-check. This is rather useless in our case since it just generates stack-probes at function entries, to ensure that (assuming a virtual memory) the stack is accessible.

- -fstack-limit-register and -fstack-limit-symbol Both are (in its current form) also quite useless, since the generated check is for 68020+ only (it uses a conditional trapcc instruction). Furthermore, -fstack-limit-register would require to reserve a register for the whole program for this purpose, and -fstack-limit-symbol compares to the value of a symbol (ie. its address, not a variable value).

So i've implemented a small patch that generates different code, if you want to try. It is for gcc-7.5.0, but should not be too hard to adapt to other versions. If you use -fstack-check, the generated code on function entry will then look like:

```
_test:
        cmp.l __StkLim.l,%sp
        jhi .L6
        jmp __StkOver
.L6:
```

But it also has a few drawbacks:
- First off, it needs a changed mintlib. This is probably always needed no matter what we try to implement. In this case, it is responsible to set the variable _StkLim to the value that should be checked, and provide the function _StkOver which is called when a stack overflow is detected.
- The code is generated in the prologue of a function. Looks like this is sometimes to early, depending on whether -fomit-frame-pointer is used or not. For example registers that need to be preserved are pushed *after* this. But this can maybe be solved by setting the limit to some reasonable value a little above the "real" stack bottom.
- i haven't found a simple way yet to also do checks for dynamic stack allocations (VLA, or use of alloca).
- it would also be nice to have the _StkOver function to report atleast the function name.
- You need to declare some functions that should not be checked with __attribute__((no_stack_limit))  (i've used that because it already exists, if needed we can also add a new attribute). That includes functions called by Supexec, functions that are callbacks for AES userdefs etc.

What do you think about it?

[stack2.patch.txt](https://github.com/freemint/freemint/files/11213672/stack2.patch.txt)

**Comment by mikrosk** [comment]
> For example registers that need to be preserved are pushed after this

Is this really a problem? I mean your routine doesn't use any registers and if the `_StkOver` does, who cares? Its purpose is to print something and exit, isn't it?

Otherwise it looks pretty non-invasive. I agree that some change to mintlib is inevitable one way or another so I'm fine with that.

**Comment by th-otto** [comment]
Yes, thats right but that means that the sp may be even lower in the real function. But i guess  you have to add some kind of safety value anyway when calculating the value for _StkLim.

**Comment by th-otto** [comment]
Here is a corresponding patch for mintlib.

[stkover.patch.txt](https://github.com/freemint/freemint/files/11225306/stkover.patch.txt)
